wallet/wallet.c

```diff
@@ -3445,26 +3445,27 @@ struct wallet_transaction *wallet_transactions_get(struct wallet *w, const tal_t
 		"  transaction_annotations a ON (a.txid = t.id) LEFT JOIN"
 		"  channels c ON (a.channel = c.id) LEFT JOIN"
 		"  channels c2 ON (t.channel_id = c2.id) "
-		"ORDER BY blockheight, txindex ASC"));
+		"ORDER BY t.blockheight, t.txindex ASC"));
 	db_query_prepared(stmt);
 
 	for (count = 0; db_step(stmt); count++) {
 		struct bitcoin_txid curtxid;
 		db_column_txid(stmt, 0, &curtxid);
 
 		/* If this is a new entry, allocate it in the array and set
-		 * the common fields (all fields from the transactions
-		 * table. */
+		 * the common fields (all fields from the transactions table. */
 		if (!bitcoin_txid_eq(&last, &curtxid)) {
 			last = curtxid;
-			tal_resize(&txs, count + 1);
-			cur = &txs[count];
+			tal_resize(&txs, tal_count(txs) + 1);
+			cur = &txs[tal_count(txs) - 1];
 			db_column_txid(stmt, 0, &cur->id);
 			cur->tx = db_column_tx(txs, stmt, 1);
 			cur->rawtx = tal_dup_arr(txs, u8, db_column_blob(stmt, 1),
 						 db_column_bytes(stmt, 1), 0);
-			cur->blockheight = db_column_int(stmt, 2);
-			cur->txindex = db_column_int(stmt, 3);
+            if (!db_column_is_null(stmt, 2))    // MSC: tx may be unconfirmed
+                cur->blockheight = db_column_int(stmt, 2);
+            if (!db_column_is_null(stmt, 3))
+                cur->txindex = db_column_int(stmt, 3);
```

Comment:
The two `if` statements should likely be combined (and an assert added): we should never have a `blockheight` without also having a `txindex`, and vice-versa.

Reply:
ok